### ISSUE TITLE
Authentication using OIDC

### ISSUE THREAD

**Issue by rizadiawan** [awslabs/kubeflow-manifests#305]
**What is the URL of the document?**
The URL to help identify the document.
https://awslabs.github.io/kubeflow-manifests/docs/deployment/vanilla/guide/#dex

**Which section(s) is the issue in?**
The sections(s) within the document that have issue in.
Dex

**What needs fixing and describe the solution you'd like?**
A clear and concise description of what the issue is.
Need guides/examples of setting up authentication with OIDC as alternative to Cognito for managing user authentication.

**Additional context**
Cognito is not supported in some AWS region like Milan, Cape Town, and Jakarta, but sometimes people have to run workload in specific region for regulatory/compliance reason.

**Comment by surajkota** [comment]
Hi @rizadiawan, thanks for creating the issue. For your usecase, there are 2 ways to go about this:
- If you plan to with set up a load balancer, you can directly configure the load balancer to authenticate using [auth-idp-oidc](https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.4/guide/ingress/annotations/#auth-idp-oidc) annotation on Ingress or
- [dex connector for OIDC based iDP](https://dexidp.io/docs/connectors/oidc/)

I havent used either of this but have experience configuring Load balancer based authentication. Would you be interested in looking into this? We can guide you based on your idP. If you are okay with using Load balancer, following are the details you will need from your idP to configure it:

1. Issuer Endpoint
2. Authorization Endpoint
3. UserInfo Endpoint
4. Token Endpoint
5. ClientId and Client Secret for your App in idP

I am asking these information based on the following docs:
- https://docs.aws.amazon.com/elasticloadbalancing/latest/application/listener-authenticate-users.html
- https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.4/guide/ingress/annotations/#auth-idp-oidc


I expect these information to be used on the ingress like:

```
apiVersion: networking.k8s.io/v1beta1
kind: Ingress
metadata:
  name: istio-ingress
  annotations:
    alb.ingress.kubernetes.io/auth-type: oidc
    alb.ingress.kubernetes.io/auth-idp-oidc: '{"Issuer":"$(oidcIssuer)","AuthorizationEndpoint":"$(oidcAuthorizationEndpoint)","TokenEndpoint":"$(oidcTokenEndpoint)","UserInfoEndpoint":"$(oidcUserInfoEndpoint)","SecretName":"$(oidcSecretName)"}'
    alb.ingress.kubernetes.io/certificate-arn: $(certArn)
    alb.ingress.kubernetes.io/listen-ports: '[{"HTTPS":443}]'
    alb.ingress.kubernetes.io/auth-scope: 'email openid profile'
    alb.ingress.kubernetes.io/scheme: internet-facing
    kubernetes.io/ingress.class: alb
spec:
  rules:
  - http:
      paths:
      - backend:
          serviceName: istio-ingressgateway
          servicePort: 80
        path: /*
```

oidcSecretName will be a K8s secret with the ClientId and Client Secret

**Comment by rizadiawan** [comment]
Hi @surajkota, thanks for reply. 

I've followed your guide above with an Auth0 app, and user is correctly presented with Auth0 login screen when opening the dashboard URL. However, upon entering correct credential, instead of logging in the user to the dashboard, the user is redirected back to local Dex login URL, for example: https://kubeflow.kubeflow.rizasaputra.com/dex/auth/local?req=q356x74qwzc7atv5fb234ecyh. 

Only after entering local Dex credential the user is taken to dashboard.

For some additional context, this is what I do:

1. Set up an Auth0 app in a domain, and obtained these details:
    -  Issuer Endpoint: https://kftest.jp.auth0.com/
    - Authorization Endpoint: https://kftest.jp.auth0.com/authorize
    - UserInfo Endpoint: https://kftest.jp.auth0.com/userinfo
    - Token Endpoint: https://kftest.jp.auth0.com/oauth/token
    - Client ID and client secret
2. Do some necessary setup:
    - Add permission to read secrets to Clusterrole aws-load-balancer-controller-role
    - Remove duplicate `aws-load-balancer-controller` service account definition from load balancer controller manifest as instructed here: https://docs.aws.amazon.com/eks/latest/userguide/aws-load-balancer-controller.html on point 5.b.ii in tab Kubernetes manifest
3. Register users on Auth0 and create corresponding profiles manually with same email https://github.com/rizadiawan/kubeflow-manifests/blob/oidc-auth/deployments/profile.yaml
4. Try login from main page: https://kubeflow.kubeflow.rizasaputra.com

Complete reference of my changes are on this commit: https://github.com/rizadiawan/kubeflow-manifests/commit/b2fb5d817640f8c96238909e51c6cb43306658d2

I've tried to also apply AWS Authservice like the Cognito deployment: `kustomize build awsconfigs/common/aws-authservice/base | kubectl apply -f -` but I still faced same problem.

Do I miss something?

**Comment by rizadiawan** [comment]
Oh, I managed to get it works. Turns out the problem is because I mixed up instructions for doing vanilla deployment, manual LB deployment, and manual Cognito deployment. Later I'll try to open a PR for docs on authenticating with ALB + idP. 

Thanks again for the guide :)

**Comment by surajkota** [comment]
Awesome, this is great! Thanks for the POC

I will resolve the issue for now but feel free to open a new PR later
